File: Application/src/main/java/com/example/android/autofillframework/app/CustomVirtualView.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.android.autofillframework.app;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Style;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.Log;
import android.util.SparseArray;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewStructure;
import android.view.autofill.AutofillManager;
import android.view.autofill.AutofillValue;
import android.widget.EditText;
import android.widget.TextView;

import com.example.android.autofillframework.R;

import java.util.ArrayList;
import java.util.Arrays;

import static com.example.android.autofillframework.CommonUtil.bundleToString;


/**
 * Custom View with virtual child views for Username/Password text fields.
 */
public class CustomVirtualView extends View {

    private static final String TAG = "CustomView";

    private static final int TOP_MARGIN = 100;
    private static final int LEFT_MARGIN = 100;
    private static final int TEXT_HEIGHT = 90;
    private static final int VERTICAL_GAP = 10;
    private static final int LINE_HEIGHT = TEXT_HEIGHT + VERTICAL_GAP;
    private static final int UNFOCUSED_COLOR = Color.BLACK;
    private static final int FOCUSED_COLOR = Color.RED;
    private static int sNextId;

    private final ArrayList<Line> mVirtualViewGroups = new ArrayList<>();
    private final SparseArray<Item> mVirtualViews = new SparseArray<>();
    private final AutofillManager mAutofillManager;

    private Line mFocusedLine;
    private Paint mTextPaint;

    private Line mUsernameLine;
    private Line mPasswordLine;

    public CustomVirtualView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mAutofillManager = context.getSystemService(AutofillManager.class);
        mTextPaint = new Paint();
        mTextPaint.setStyle(Style.FILL);
        mTextPaint.setTextSize(TEXT_HEIGHT);
        mUsernameLine = addLine("usernameField", context.getString(R.string.username_label),
                new String[]{View.AUTOFILL_HINT_USERNAME}, "         ", true);
        mPasswordLine = addLine("passwordField", context.getString(R.string.password_label),
                new String[]{View.AUTOFILL_HINT_PASSWORD}, "         ", false);
    }

    @Override
    public void autofill(SparseArray<AutofillValue> values) {
        // User has just selected a Dataset from the list of autofill suggestions.
        // The Dataset is comprised of a list of AutofillValues, with each AutofillValue meant
        // to fill a specific autofillable view. Now we have to update the UI based on the
        // AutofillValues in the list.
        Log.d(TAG, "autoFill(): " + values);
        for (int i = 0; i < values.size(); i++) {
            int id = values.keyAt(i);
            AutofillValue value = values.valueAt(i);
            Item item = mVirtualViews.get(id);
            if (item != null && item.editable) {
                // Set the item's text to the text wrapped in the AutofillValue.
                item.text = value.getTextValue();
            } else if (item == null) {
                Log.w(TAG, "No item for id " + id);
            } else {
                Log.w(TAG, "Item for id " + id + " is not editable: " + item);
            }
        }
        postInvalidate();
    }


    @Override
    public void onProvideAutofillVirtualStructure(ViewStructure structure, int flags) {
        // Build a ViewStructure that will get passed to the AutofillService by the framework
        // when it is time to find autofill suggestions.
        structure.setClassName(getClass().getName());
        int childrenSize = mVirtualViews.size();
        Log.d(TAG, "onProvideAutofillVirtualStructure(): flags = " + flags + ", items = "
                + childrenSize + ", extras: " + bundleToString(structure.getExtras()));
        int index = structure.addChildCount(childrenSize);
        // Traverse through the view hierarchy, including virtual child views. For each view, we
        // need to set the relevant autofill metadata and add it to the ViewStructure.
        for (int i = 0; i < childrenSize; i++) {
            Item item = mVirtualViews.valueAt(i);
            Log.d(TAG, "Adding new child at index " + index + ": " + item);
            ViewStructure child = structure.newChild(index);
            child.setAutofillId(structure.getAutofillId(), item.id);
            child.setAutofillHints(item.hints);
            child.setAutofillType(item.type);
            child.setDataIsSensitive(!item.sanitized);
            child.setText(item.text);
            child.setAutofillValue(AutofillValue.forText(item.text));
            child.setFocused(item.focused);
            child.setId(item.id, getContext().getPackageName(), null, item.line.idEntry);
            child.setClassName(item.getClassName());
            index++;
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        Log.d(TAG, "onDraw: " + mVirtualViewGroups.size() + " lines; canvas:" + canvas);
        float x;
        float y = TOP_MARGIN + LINE_HEIGHT;
        for (int i = 0; i < mVirtualViewGroups.size(); i++) {
            x = LEFT_MARGIN;
            Line line = mVirtualViewGroups.get(i);
            Log.v(TAG, "Drawing '" + line + "' at " + x + "x" + y);
            mTextPaint.setColor(line.fieldTextItem.focused ? FOCUSED_COLOR : UNFOCUSED_COLOR);
            String readOnlyText = line.labelItem.text + ":  [";
            String writeText = line.fieldTextItem.text + "]";
            // Paints the label first...
            canvas.drawText(readOnlyText, x, y, mTextPaint);
            // ...then paints the edit text and sets the proper boundary
            float deltaX = mTextPaint.measureText(readOnlyText);
            x += deltaX;
            line.bounds.set((int) x, (int) (y - LINE_HEIGHT),
                    (int) (x + mTextPaint.measureText(writeText)), (int) y);
            Log.d(TAG, "setBounds(" + x + ", " + y + "): " + line.bounds);
            canvas.drawText(writeText, x, y, mTextPaint);
            y += LINE_HEIGHT;
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int y = (int) event.getY();
        Log.d(TAG, "Touched: y=" + y + ", range=" + LINE_HEIGHT + ", top=" + TOP_MARGIN);
        int lowerY = TOP_MARGIN;
        int upperY = -1;
        for (int i = 0; i < mVirtualViewGroups.size(); i++) {
            upperY = lowerY + LINE_HEIGHT;
            Line line = mVirtualViewGroups.get(i);
            Log.d(TAG, "Line " + i + " ranges from " + lowerY + " to " + upperY);
            if (lowerY <= y && y <= upperY) {
                if (mFocusedLine != null) {
                    Log.d(TAG, "Removing focus from " + mFocusedLine);
                    mFocusedLine.changeFocus(false);
                }
                Log.d(TAG, "Changing focus to " + line);
                mFocusedLine = line;
                mFocusedLine.changeFocus(true);
                invalidate();
                break;
            }
            lowerY += LINE_HEIGHT;
        }
        return super.onTouchEvent(event);
    }

    public CharSequence getUsernameText() {
        return mUsernameLine.fieldTextItem.text;
    }

    public CharSequence getPasswordText() {
        return mPasswordLine.fieldTextItem.text;
    }

    public void resetFields() {
        mUsernameLine.reset();
        mPasswordLine.reset();
        postInvalidate();
    }

    private Line addLine(String idEntry, String label, String[] hints, String text,
            boolean sanitized) {
        Line line = new Line(idEntry, label, hints, text, sanitized);
        mVirtualViewGroups.add(line);
        mVirtualViews.put(line.labelItem.id, line.labelItem);
        mVirtualViews.put(line.fieldTextItem.id, line.fieldTextItem);
        return line;
    }

    private static final class Item {
        private final Line line;
        private final int id;
        private final boolean editable;
        private final boolean sanitized;
        private final String[] hints;
        private final int type;
        private CharSequence text;
        private boolean focused = false;

        Item(Line line, int id, String[] hints, int type, CharSequence text, boolean editable,
                boolean sanitized) {
            this.line = line;
            this.id = id;
            this.text = text;
            this.editable = editable;
            this.sanitized = sanitized;
            this.hints = hints;
            this.type = type;
        }

        @Override
        public String toString() {
            return id + ": " + text + (editable ? " (editable)" : " (read-only)"
                    + (sanitized ? " (sanitized)" : " (sensitive"));
        }

        public String getClassName() {
            return editable ? EditText.class.getName() : TextView.class.getName();
        }
    }

    private final class Line {

        // Boundaries of the text field, relative to the CustomView
        final Rect bounds = new Rect();
        private Item labelItem;
        private Item fieldTextItem;
        private String idEntry;

        private Line(String idEntry, String label, String[] hints, String text, boolean sanitized) {
            this.idEntry = idEntry;
            this.labelItem = new Item(this, ++sNextId, null, AUTOFILL_TYPE_NONE, label,
                    false, true);
            this.fieldTextItem = new Item(this, ++sNextId, hints, AUTOFILL_TYPE_TEXT, text,
                    true, sanitized);
        }

        void changeFocus(boolean focused) {
            fieldTextItem.focused = focused;
            if (focused) {
                Rect absBounds = getAbsCoordinates();
                Log.d(TAG, "focus gained on " + fieldTextItem.id + "; absBounds=" + absBounds);
                mAutofillManager.notifyViewEntered(CustomVirtualView.this, fieldTextItem.id,
                        absBounds);
            } else {
                Log.d(TAG, "focus lost on " + fieldTextItem.id);
                mAutofillManager.notifyViewExited(CustomVirtualView.this, fieldTextItem.id);
            }
        }

        private Rect getAbsCoordinates() {
            // Must offset the boundaries so they're relative to the CustomView.
            int offset[] = new int[2];
            getLocationOnScreen(offset);
            Rect absBounds = new Rect(bounds.left + offset[0],
                    bounds.top + offset[1],
                    bounds.right + offset[0], bounds.bottom + offset[1]);
            Log.v(TAG, "getAbsCoordinates() for " + fieldTextItem.id + ": bounds=" + bounds
                    + " offset: " + Arrays.toString(offset) + " absBounds: " + absBounds);
            return absBounds;
        }

        public void reset() {
            fieldTextItem.text = "        ";
        }

        @Override
        public String toString() {
            return "Label: " + labelItem + " Text: " + fieldTextItem + " Focused: " +
                    fieldTextItem.focused;
        }
    }
}

File: kotlinApp/Application/src/main/java/com/example/android/autofillframework/app/CustomVirtualView.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.android.autofillframework.app

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.Style
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.util.SparseArray
import android.view.MotionEvent
import android.view.View
import android.view.ViewStructure
import android.view.autofill.AutofillManager
import android.view.autofill.AutofillValue
import android.widget.EditText
import android.widget.TextView
import com.example.android.autofillframework.CommonUtil.TAG
import com.example.android.autofillframework.CommonUtil.bundleToString
import com.example.android.autofillframework.R
import java.util.Arrays


/**
 * Custom View with virtual child views for Username/Password text fields.
 */
class CustomVirtualView(context: Context, attrs: AttributeSet) : View(context, attrs) {

    val usernameText: CharSequence
        get() = usernameLine.fieldTextItem.text
    val passwordText: CharSequence
        get() = passwordLine.fieldTextItem.text
    private var nextId: Int = 0
    private val lines = ArrayList<Line>()
    private val items = SparseArray<Item>()
    private val autofillManager = context.getSystemService(AutofillManager::class.java)
    private var focusedLine: Line? = null
    private val textHeight = 90
    private val textPaint = Paint().apply {
        style = Style.FILL
        textSize = textHeight.toFloat()
    }
    private val topMargin = 100
    private val leftMargin = 100
    private val verticalGap = 10
    private val lineLength = textHeight + verticalGap
    private val focusedColor = Color.RED
    private val unfocusedColor = Color.BLACK
    private val usernameLine = addLine("usernameField", context.getString(R.string.username_label),
            arrayOf(View.AUTOFILL_HINT_USERNAME), "         ", true)
    private val passwordLine = addLine("passwordField", context.getString(R.string.password_label),
            arrayOf(View.AUTOFILL_HINT_PASSWORD), "         ", false)

    override fun autofill(values: SparseArray<AutofillValue>) {
        // User has just selected a Dataset from the list of autofill suggestions.
        // The Dataset is comprised of a list of AutofillValues, with each AutofillValue meant
        // to fill a specific autofillable view. Now we have to update the UI based on the
        // AutofillValues in the list.
        Log.d(TAG, "autofill(): " + values)
        for (i in 0 until values.size()) {
            val id = values.keyAt(i)
            val value = values.valueAt(i)
            items[id]?.apply {
                if (editable) {
                    // Set the item's text to the text wrapped in the AutofillValue.
                    text = value.textValue
                } else {
                    Log.w(TAG, "Item for autofillId $id is not editable: $this")
                }
            }
        }
        postInvalidate()
    }

    override fun onProvideAutofillVirtualStructure(structure: ViewStructure, flags: Int) {
        // Build a ViewStructure to pack in AutoFillService requests.
        structure.setClassName(javaClass.name)
        val childrenSize = items.size()
        Log.d(TAG, "onProvideAutofillVirtualStructure(): flags = " + flags + ", items = "
                + childrenSize + ", extras: " + bundleToString(structure.extras))
        var index = structure.addChildCount(childrenSize)
        for (i in 0 until childrenSize) {
            val item = items.valueAt(i)
            Log.d(TAG, "Adding new child at index $index: $item")
            structure.newChild(index).apply {
                setAutofillId(structure.autofillId, item.id)
                setAutofillHints(item.hints)
                setAutofillType(item.type)
                setDataIsSensitive(!item.sanitized)
                setAutofillValue(AutofillValue.forText(item.text))
                setFocused(item.focused)
                setId(item.id, context.packageName, null, item.line.idEntry)
                setClassName(item.className)
            }
            index++
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        Log.d(TAG, "onDraw: " + lines.size + " lines; canvas:" + canvas)
        var x: Float
        var y = (topMargin + lineLength).toFloat()

        lines.forEach {
            x = leftMargin.toFloat()
            Log.v(TAG, "Drawing $it at x=$x, y=$y")
            textPaint.color = if (it.fieldTextItem.focused) focusedColor else unfocusedColor
            val readOnlyText = it.labelItem.text.toString() + ":  ["
            val writeText = it.fieldTextItem.text.toString() + "]"
            // Paints the label first...
            canvas.drawText(readOnlyText, x, y, textPaint)
            // ...then paints the edit text and sets the proper boundary
            val deltaX = textPaint.measureText(readOnlyText)
            x += deltaX
            it.bounds.set(x.toInt(), (y - lineLength).toInt(),
                    (x + textPaint.measureText(writeText)).toInt(), y.toInt())
            Log.d(TAG, "setBounds(" + x + ", " + y + "): " + it.bounds)
            canvas.drawText(writeText, x, y, textPaint)
            y += lineLength.toFloat()
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val y = event.y.toInt()
        Log.d(TAG, "Touched: y=$y, range=$lineLength, top=$topMargin")
        var lowerY = topMargin
        var upperY = -1
        for (line in lines) {
            upperY = lowerY + lineLength
            Log.d(TAG, "Line $line ranges from $lowerY to $upperY")
            if (y in lowerY..upperY) {
                Log.d(TAG, "Removing focus from " + focusedLine)
                focusedLine?.changeFocus(false)
                Log.d(TAG, "Changing focus to " + line)
                focusedLine = line.apply { changeFocus(true) }
                invalidate()
                break
            }
            lowerY += lineLength
        }
        return super.onTouchEvent(event)
    }

    fun resetFields() {
        usernameLine.reset()
        passwordLine.reset()
        postInvalidate()
    }

    private fun addLine(idEntry: String, label: String, hints: Array<String>, text: String,
            sanitized: Boolean) = Line(idEntry, label, hints, text, sanitized).also {
        lines.add(it)
        items.apply {
            put(it.labelItem.id, it.labelItem)
            put(it.fieldTextItem.id, it.fieldTextItem)
        }
    }

    private inner class Item internal constructor(
            val line: Line,
            val id: Int,
            val hints: Array<String>?,
            val type: Int, var text: CharSequence, val editable: Boolean,
            val sanitized: Boolean) {
        var focused = false

        override fun toString(): String {
            return id.toString() + ": " + text + if (editable)
                " (editable)"
            else
                " (read-only)" + if (sanitized) " (sanitized)" else " (sensitive"
        }

        val className: String
            get() = if (editable) EditText::class.java.name else TextView::class.java.name
    }

    private inner class Line constructor(val idEntry: String, label: String, hints: Array<String>,
            text: String, sanitized: Boolean) {

        // Boundaries of the text field, relative to the CustomView
        internal val bounds = Rect()
        var labelItem: Item = Item(this, ++nextId, null, View.AUTOFILL_TYPE_NONE, label, false, true)
        var fieldTextItem: Item = Item(this, ++nextId, hints, View.AUTOFILL_TYPE_TEXT, text, true, sanitized)

        internal fun changeFocus(focused: Boolean) {
            fieldTextItem.focused = focused
            if (focused) {
                val absBounds = absCoordinates
                Log.d(TAG, "focus gained on " + fieldTextItem.id + "; absBounds=" + absBounds)
                autofillManager.notifyViewEntered(this@CustomVirtualView, fieldTextItem.id, absBounds)
            } else {
                Log.d(TAG, "focus lost on " + fieldTextItem.id)
                autofillManager.notifyViewExited(this@CustomVirtualView, fieldTextItem.id)
            }
        }

        private val absCoordinates: Rect
                // Must offset the boundaries so they're relative to the CustomView.
            get() {
                val offset = IntArray(2)
                getLocationOnScreen(offset)
                val absBounds = Rect(bounds.left + offset[0],
                        bounds.top + offset[1],
                        bounds.right + offset[0], bounds.bottom + offset[1])
                Log.v(TAG, "absCoordinates for " + fieldTextItem.id + ": bounds=" + bounds
                        + " offset: " + Arrays.toString(offset) + " absBounds: " + absBounds)
                return absBounds
            }

        fun reset() {
            fieldTextItem.text = "        "
        }

        override fun toString(): String {
            return "Label: " + labelItem + " Text: " + fieldTextItem + " Focused: " +
                    fieldTextItem.focused
        }
    }
}